### DISCUSSION TITLE
Dynamic completion codes for CloudResearch (turkPrime) mTurk

### DISCUSSION THREAD

**Discussion by dreamerie** [jspsych/jsPsych#710] [Q&A] [answered]
Hi all, I have been using cloudresearch/turkprime to collect data from mTurk workers.I have been using the fixed completion code option so that workers who enter the code (displayed at the end of the study) are automatically approved.

However, I would like to use the dynamic completion code as this is more secure:
https://www.cloudresearch.com/resources/blog/qualtrics-and-other-platforms-mechanical-turk-integration-with-cloudresearch/

Their guide only specifies how to do this with Qualtrics, but I imagine it can be integrated with jspsych code as well.
It seems to involve displaying an iframe but I am not sure how to achieve this (on_finish in jsPsych.init?)

**Comment by jodeleeuw** [comment, marked as answer]
With the caveat that I haven't tried this yet, I think you could accomplish it with the `html-keyboard-response` plugin as the final trial of the experiment:

```js
var dynamic_code_trial = {
	type: 'html-keyboard-response',
	choices: jsPsych.NO_KEYS, // prevent the subject from continuing after this point
	stimulus: '<iframe src="https://www.turkprime.com/TurkPrime/TakeLaunchedSurvey/DynamicKey" width="100%" height="200"></iframe>'
}
```

Please let us know if this works!

**Reply by dreamerie** [reply]
It worked! Thanks so much Josh!

For anyone who has the same question, I also had to move saveData to be in here instead of at the end of the study:

```
var dynamic_code_trial = {
	type: 'html-keyboard-response',
	choices: jsPsych.NO_KEYS,
	stimulus: '<iframe src="https://www.turkprime.com/TurkPrime/TakeLaunchedSurvey/DynamicKey" width="100%" height="200"></iframe>',
  on_load: function () {
    saveData(id + ".csv", jsPsych.data.get().csv());
  }
}
```
